### PR TITLE
Event emitter

### PR DESCRIPTION
**Pull request by Hotell**
implements `EventEmitter` functionality like angular 2 does
- beware it is not a real Rx Subject, just abstraction

cc @dmorosinotto

### CODE REVIEW

**Comment by dmorosinotto** [comment]
Hi @Hotell at the end you decided to go for the "class path" implementing a mock of the RxJS Observable for the EventEmitter... 
You don't think that the "interface approach" that I [suggested](https://github.com/dmorosinotto/ng-metadata-quickstart/blob/EventEmitter/app/components/ng-metadata_core_patch.ts) to you is sufficient to have type-safety and similarity to the NG2 code?

I still see missing something at the consumer side (parent comp) to have typecheck for the code that you use to handle the event. 
I think that having something like the `interface EventHandler<T>` that I propouse to you, can help to check the [implementation code](https://github.com/dmorosinotto/ng-metadata-quickstart/blob/EventEmitter/app/components/my-app.component.ts#L30-L32) and avoid some error... (I know that it's not an NG2 feature, but ngMetadata at the end run over NG1)

I'm really scratching my head thinking about it, and reading the code that you wrote at the end to accomplish the emit method, let's me digest it, and give my final thought on it in the weekend, because right now I'm under the gun on a project release and I can't take the right time to try it and evaluate/give my feedback. 

Obviusly if you wan't it, if you have just stated that this is the way that have to be, then let's go for it!
 
Let's me think on it and maybe we can chat here or in slack in the weekend

File: playground/app/components/todo-item.component.ts [hunk 1]
```diff
@@ -12,9 +12,19 @@ export class TodoItemCmp implements OnChanges{
 
   @Input('<') todo: TodoModel;
   @Input('<') idx: number;
-  @Output() onDone: ( todo: {todo:TodoModel} )=>void;
+  // this is old way how to bindings are created, will be removed in 2.0
+  // @Output() onDone: ( todo: {todo:TodoModel} )=>void;
+  @Output() onDone: EventEmitter<TodoModel>;
```
Comment (dmorosinotto) [inline playground/app/components/todo-item.component.ts]: maybe here you are missing a `new` on the onDone initialization

File: playground/app/components/todo-item.component.ts [hunk 2]
```diff
@@ -12,9 +12,19 @@ export class TodoItemCmp implements OnChanges{
...
+  @Output() emitTest = new EventEmitter<string>();
 
-  ngOnInit(){}
+  ngOnInit(){
+
+    const dispose = this.onDone.subscribe( (value)=> {
+      console.log( `onDone emitted value: ${value}` );
```
Comment (dmorosinotto) [inline playground/app/components/todo-item.component.ts]: This code is only to demostrate that you implemented the subscribe method on the EventEmitter conforming to RxJS, but I still don't understand why you want it on the "client side" why a component has to register and subscribe to the event that emit itself?
In NG2 I see the value of this approach to handle events but on the *parent code* (comp that handle the event using subscribe + RxJS operators) but not here... 
If I correctly understood your code, even if you've created the EventEmitter class I can use it only on this side!? Or I'm missing something / don't undestood it?

Comment (Hotell) [inline playground/app/components/todo-item.component.ts]: you don't have to subscribe, that's just showcase that subscribe works for EventEmmiter, and its registered "observer/callback" will be called when it's emitted via `this.emit('booo')`

Comment (Hotell) [inline playground/app/components/todo-item.component.ts]: I added the `subscribe` but it is not intended to use, but If you need to trigger some other sideffect during `emit` you can, that's all, no rocket science :)

File: playground/app/components/todo-item.component.ts (hunk 1 above)
Comment (Hotell) [inline playground/app/components/todo-item.component.ts]: In contrary, you can do it your way and my way, both works ;) interface of new instance, everybody wins! ;)

**Comment by Hotell** [comment]
any chance to re-review @dmorosinotto ? otherwise I'll merge it. need to do this before going to 2.0

**Comment by dmorosinotto** [comment]
Hi @Hotell sorry for the delay, but it took me a while to understand you code, expecially because I was digging a bit on the code that take the binding and create the _ddo.
I'm not sure of fully undestand it, maybe a day you can explain/tutor me the internals of ngMetadata, but leaving this for other discussion... 
I tried your solution and see that it works with the double approach (class and interface) this it GREAT! :+1:

The only part that maybe I'd change is the add of an interface EventHandler<T> 
```typescript
export interface EventHandler<T> {
  ($event: T): void;
}
```
that enable a more explicit typecheck for generatorOrNext function passed to subscribe, but even more important can be used on the parent component (consumer side) to delcare the function handling the event with something like:
```typescript
outputWithEmitReciver: EventHandler<string> = ( sentence) => {
    console.log(sentence):  //HERE sentence is a string
}

markAsDone: EventHandler<TodoModel> = (todo) => {
    //HERE todo is a TodoModel
   ...
}
```

I don't know your opionion on that, if you are ok introducing an Interface that doesn't exist in the NG2 nomenclature, but in my opinion it allows to define in a more explicit way that a function is an handler for somekind of event emitted by a child comp and sharing the type T of the payload you can have intellisence on both side for $event properties.

If you think that EventHandler can be usefull I'll make a PR, or anyway all your work LGTM and so merge it when you want!

Thanks for all you do Martin, I really appreciate!

**Pull request merged**